SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/SpatialVirtualWorkerTranslationManager.h"
+#include "EngineClasses/SpatialVirtualWorkerTranslator.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialReceiver.h"
+#include "SpatialConstants.h"
+#include "Utils/SchemaUtils.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialVirtualWorkerTranslationManager);
+
+SpatialVirtualWorkerTranslationManager::SpatialVirtualWorkerTranslationManager(
+	USpatialReceiver* InReceiver,
+	USpatialWorkerConnection* InConnection,
+	SpatialVirtualWorkerTranslator* InTranslator)
+	: Receiver(InReceiver)
+	, Connection(InConnection)
+	, Translator(InTranslator)
+	, bWorkerEntityQueryInFlight(false)
+{}
+
+void SpatialVirtualWorkerTranslationManager::AddVirtualWorkerIds(const TSet<VirtualWorkerId>& InVirtualWorkerIds)
+{
+	// Currently, this should only be called once on startup. In the future we may allow for more
+	// flexibility. 
+	check(UnassignedVirtualWorkers.IsEmpty());
+	for (VirtualWorkerId VirtualWorkerId : InVirtualWorkerIds)
+	{
+		UnassignedVirtualWorkers.Enqueue(VirtualWorkerId);
+	}
+}
+
+void SpatialVirtualWorkerTranslationManager::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
+{
+	const bool bAuthoritative = AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE;
+
+	if (!bAuthoritative)
+	{
+		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Error, TEXT("Lost authority over the translation mapping. This is not supported."));
+		return;
+	}
+
+	check(AuthOp.component_id == SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID);
```

Comment:
NIT Could maybe just put this check at the start of the method

Reply:
ok